be/src/exec/vectorized/sorting/merge_column.cpp

```diff
@@ -0,0 +1,229 @@
+                   Permutation* perm)
+            : ColumnVisitorAdapter(this),
+              _sort_order(desc.sort_order),
+              _null_first(desc.null_first),
+              _left_col(left_col),
+              _right_col(right_col),
+              _equal_ranges(equal_range),
+              _perm(perm) {}
+
+    template <class ColumnType>
+    void do_merge() {
+        std::vector<EqualRange> next_ranges;
+        next_ranges.reserve(_equal_ranges->size());
+        auto left_col = down_cast<const ColumnType*>(_left_col);
+        auto right_col = down_cast<const ColumnType*>(_right_col);
+
+        // Iterate each equal-range
+        for (auto equal_range : *_equal_ranges) {
+            size_t lhs = equal_range.left_range.first;
+            size_t rhs = equal_range.right_range.first;
+            size_t lhs_end = equal_range.left_range.second;
+            size_t rhs_end = equal_range.right_range.second;
+            size_t output_index = lhs + rhs;
+
+            // Merge rows in the equal-range
+            while (lhs < lhs_end || rhs < rhs_end) {
+                auto left_range = fetch_left(left_col, lhs, lhs_end);
+                auto right_range = fetch_right(right_col, rhs, rhs_end);
+
+                // TODO: optimize the compare
+                int x = 0;
+                if (lhs < lhs_end && rhs < rhs_end) {
+                    x = left_col->compare_at(left_range.first, right_range.first, *right_col, _null_first);
+                    x *= _sort_order;
+                } else if (lhs < lhs_end) {
+                    x = -1;
+                } else if (rhs < rhs_end) {
+                    x = 1;
+                }
+
+                if (x <= 0) {
+                    lhs = left_range.second;
+                    for (size_t i = left_range.first; i < left_range.second; i++) {
+                        (*_perm)[output_index++] = PermutationItem(kLeftIndex, i);
+                    }
+                }
+                if (x >= 0) {
+                    rhs = right_range.second;
+                    for (size_t i = right_range.first; i < right_range.second; i++) {
+                        (*_perm)[output_index++] = PermutationItem(kRightIndex, i);
+                    }
+                }
+                if (x == 0) {
+                    next_ranges.emplace_back(left_range, right_range);
+                }
+            }
+
+            DCHECK_EQ(lhs, lhs_end);
+            DCHECK_EQ(rhs, rhs_end);
+        }
+
+        _equal_ranges->swap(next_ranges);
+    }
+
+    template <class ColumnType>
+    Status do_visit(const ColumnType&) {
+        do_merge<ColumnType>();
+        return Status::OK();
+    }
+
+    template <class ColumnType>
+    EqualRange::Range fetch_left(const ColumnType* left_col, size_t lhs, size_t lhs_end) {
+        uint32_t first = lhs;
+        uint32_t last = lhs + 1;
+        while (last < lhs_end && left_col->compare_at(last - 1, last, *left_col, _null_first) == 0) {
```

Comment:
If we only need to check the equal for the same column. I think we could implement a new method equal_at to replace compare_at , then there should be less branch miss.

```
struct SorterComparator {
    static int compare(const T& lhs, const T& rhs) {
        if (lhs == rhs) {
            return 0;
        } else if (lhs < rhs) {
            return -1;
        } else {
            return 1;
        }
    }
};
```

Reply:
done. I will use the `operator ==` to replace `compare`.